File: src/insertion.sort.cpp
template <typename T> void insertionSort(T *arr, int n, bool ascending = true) {
  for (int i = 1; i < n; i++) {
    T   key = arr[i];
    int j = i - 1;
    if (ascending) {
      while (j >= 0 && arr[j] > key) {
        arr[j + 1] = arr[j];
        j--;
      }
    } else {
      while (j >= 0 && arr[j] < key) {
        arr[j + 1] = arr[j];
        j--;
      }
    }
    arr[j + 1] = key;
  }
}

File: frontend/struct/src/functions/sortings/insertionSort.ts
import { SortStep } from "@/interfaces/interface";

export const insertionSortStep = (
  arr: number[],
  step: SortStep,
  ascending: boolean = true,
): { sortedArray: number[]; step: SortStep } => {
  const sortedArray = [...arr];
  let { i, j } = step;

  if (i >= sortedArray.length) {
    return { sortedArray, step: { i, j } };
  }

  const key = sortedArray[i];
  j = i - 1;

  if (ascending) {
    while (j >= 0 && sortedArray[j] > key) {
      sortedArray[j + 1] = sortedArray[j];
      j--;
    }
  } else {
    while (j >= 0 && sortedArray[j] < key) {
      sortedArray[j + 1] = sortedArray[j];
      j--;
    }
  }
  sortedArray[j + 1] = key;

  i++;

  return { sortedArray, step: { i, j } };
};
